src/django/api/views/boundary.py

```diff
@@ -0,0 +1,69 @@
+from django.http import Http404
+from django.db.models import Prefetch
+
+from rest_framework.response import Response
+from rest_framework.exceptions import APIException
+from rest_framework.views import APIView
+from rest_framework.permissions import IsAuthenticated
+
+from ..models.user import Roles
+from ..models.boundary import Boundary
+from ..models.utility import Utility
+from ..models.submission import Submission
+
+from ..serializers.boundary import BoundaryDetailSerializer, BoundaryListSerializer
+
+
+def get_boundary_queryset_for_user(user):
+    if user.role == Roles.CONTRIBUTOR:
+        return Boundary.objects.filter(utility__in=user.utilities.all())
+
+    if user.role == Roles.VALIDATOR:
+        return Boundary.objects.filter(utility__state__in=user.states.all())
+
+    if user.role == Roles.ADMINISTRATOR:
+        return Boundary.objects.all()
+
+    raise RuntimeError('Invalid role: {}'.format(user.role))
+
+
+class BoundaryListView(APIView):
+    permission_classes = [IsAuthenticated]
+
+    def get(self, request, format=None):
+        boundaries = get_boundary_queryset_for_user(request.user)
+        boundaries = boundaries.select_related('utility')
+
+        requested_utilities = request.GET.get('utilities')
+        if requested_utilities:
+            for utility_id in requested_utilities:
+                if not request.user.has_access_to_utility(
+                    Utility.objects.get(pk=utility_id)
+                ):
+                    raise APIException('Cannot access this utility', 401)
+
+            boundaries = boundaries.filter(utility__in=requested_utilities)
+
+        return Response(BoundaryListSerializer(boundaries, many=True).data)
+
+
+class BoundaryDetailView(APIView):
+    permission_classes = [IsAuthenticated]
+
+    def get(self, request, id, format=None):
+        boundary = get_boundary_queryset_for_user(request.user)
+        boundary = boundary.select_related('utility')
+        boundary = boundary.prefetch_related(
+            Prefetch(
+                'submissions',
+                queryset=Submission.objects.select_related('review').prefetch_related(
```

Comment:
On Slack you asked if we could make Review -> Submission one-to-one. If we make that change, that will change this clause a little bit.

Reply:
It turns out that review to submission is one to one and this seems to be working correctly (or at least it's returning the right data. It might be still making too many queries).

My understanding is that Submissions and Annotations have to be prefetched because they are many to one, but reviews and utilities can be selected at the same time as their parents because they are one to one.

Reply:
Oops, I was looking at the model defs on develop. Got it now. 👍🏻